### PR TITLE
[RISCV] Generalize (sub zext, zext) -> (sext (sub zext, zext)) to add

### PR DESCRIPTION
**Pull request by lukel97**
This generalizes the combine added in #82455 to other binary ops, beginning with adds in this patch.

Because the two zext operands are always +ve when treated as signed, and we don't get any overflow since the add is carried out in at least N * 2 bits of the narrow type, the result of the add will always be +ve. So we can use a zext for the outer extend, unlike sub which may produce a -ve result from two +ve operands.

Although we could still use sext for add, I plan to add support for other binary ops like mul in a later patch, but mul requires zext to be correct (because the maximum value will take up the full N * 2 bits). So I've opted to use zext here too for consistency.

Alive2 proof: https://alive2.llvm.org/ce/z/PRNsUM

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -12899,6 +12899,56 @@ static SDValue transformAddImmMulImm(SDNode *N, SelectionDAG &DAG,
...
+  Src0 = DAG.getNode(ISD::ZERO_EXTEND, SDLoc(Src0), NarrowVT, Src0);
+  Src1 = DAG.getNode(ISD::ZERO_EXTEND, SDLoc(Src1), NarrowVT, Src1);
+
+  // Src0 and Src1 are zero extended, so they're always +ve if signed.
+  //
+  // sub can produce a -ve from two +ve operands, so it needs sign
+  // extended. Other nodes produce a +ve from two +ve operands, so zero extend
+  // instead.
+  unsigned OuterExtend =
+      N->getOpcode() == ISD::SUB ? ISD::SIGN_EXTEND : ISD::ZERO_EXTEND;
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:12945]: If we are going to extend this combination to more opcodes, we can make `OuterExtend` a parameter.

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Comment by sun-jacobi** [comment]
Alive2 proof without `--disable-undef-input`: https://alive2.llvm.org/ce/z/CkgjUj

**Review by sun-jacobi — COMMENTED** [review]
LGTM

**Review by preames — APPROVED** [review]
LGTM as well.

Minor, but what is the "ve" abbreviation you use everywhere?  You should probably spell that out.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -12899,6 +12899,56 @@ static SDValue transformAddImmMulImm(SDNode *N, SelectionDAG &DAG,
...
+// where the sum of the extend widths match, and the the range of the bin op
+// fits inside the width of the narrower bin op. (For profitability on rvv, we
+// use a power of two for both inner and outer extend.)
+//
+// TODO: Extend this to other binary ops
+static SDValue combineBinOpOfZExt(SDNode *N, SelectionDAG &DAG,
+                                  const RISCVSubtarget &Subtarget) {
+
+  EVT VT = N->getValueType(0);
+  if (!VT.isVector() || !Subtarget.getTargetLowering()->isTypeLegal(VT))
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:12914]: You can use DAG.getTargetLowering() instead and drop the Subtarget.

**Comment by topperc** [comment]
> LGTM as well.
> 
> Minor, but what is the "ve" abbreviation you use everywhere? You should probably spell that out.

I think +ve means positive and -ve means negative. https://en.wiktionary.org/wiki/%2Bve

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 1 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:12945]: Out of all the binary ops we can do this transform with it looks like sub is the only one that actually requires sign_extend: https://alive2.llvm.org/ce/z/PEPb9j which was why I was just handling it as a specific case

**Pull request merged**